Частичное решение задание 10

TatyanaV24 opened the pull request:
Дмитрий, добрый день.
Застопорилась на задаче с таблицей.
Я решила создавать строки, при помощи структур и классов.
А потом все объеденить в массив, как в уроке.
Думаю я вновь перемудрила.
Первоначальный вариант таблицы(до того что сейчас), создавал таблицу и разбивал на группы, но не удавалось добавить особенности строкам. Я смотрела видео-уроки, и читала как создавать разные строки в одной таблице и там использовали UITableViewController, а не UITableView.
Мне очень нужен Ваш дельный совет, т.к. не могу понять как разные строки свести в одну таблицу и как это прописать в коде.
По условию задачи у меня есть три вариант строк:
1. Картинка+лейбл+свитч
2.Картинка+лейбл+лейбл
3.Картинка+лейбл
Буду ждать Ваш ответ по этому вопросу.

TatyanaV24 committed:
Частичное решение задание 10

Дмитрий, добрый день.
Застопорилась на задаче с таблицей.
Я решила создавать строки, при помощи структур и классов.
А потом все объеденить в массив, как в уроке.
Думаю я вновь перемудрила.
Первоначальный вариант таблицы(до того что сейчас), создавал таблицу и разбивал на группы, но не удавалось добавить особенности строкам. Я смотрела видео-уроки, и читала как создавать разные строки в одной таблице и там использовали UITableViewController, а не UITableView.
Мне очень нужен Ваш дельный совет, т.к. не могу понять как разные строки свести в одну таблицу и как это прописать в коде.
По условию задачи у меня есть три вариант строк:
1. Картинка+лейбл+свитч
2.Картинка+лейбл+лейбл
3.Картинка+лейбл
Буду ждать Ваш ответ по этому вопросу.

ios-review commented on HW10/HW10/Model/Model.swift:
```diff
@@ -0,0 +1,35 @@
+struct ThirdCell {
+    var iconImage: String?
+    var tuning: String?
+}
```
Использовать одну модель данных Setting и одну ячейку, которая может отрисовать любой вид Setting.

```
struct Setting {
    let title: String
    let icon: UIImage?
    let isSwitchHidden: Bool // здесь может быть название более точно описывающее назначение этого переключателя, но в к учебных целях ок)
    
    init(title: String, icon: UIImage?, isSwitchHidden: Bool = true)
}
```

ios-review commented on HW10/HW10/NameTableViewCell.swift:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+    }
+}
```
Одна ячейка с разными состояниями подвью в зависимости от переданной модели данных.

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -0,0 +1,62 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        
+        let cell = tableView.dequeueReusableCell(withIdentifier: "FirstCell") as! FirstTableViewCell
+        cell.firstCell = cont[indexPath.row]
```
На этой строчке компилятор выдает ошибку.
Проект запустить нельзя.

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -0,0 +1,62 @@
+            Content(number: "", names: ["cellTable10","cellTable11","cellTable12","cellTable3","cellTable4"])
+        ]
+    }
+}
```
```
struct Setting {
}
struct SettingSection {
    let title: String
    let settings: [Setting]
}

class SettingsViewController: UIViewController {
    let settingSections = [SettingSection(title: "1", settings: [Setting()])]
}

```

ios-review reviewed (changes requested):
Добавил подсказки по модели данных и ячейкам

TatyanaV24 committed:
HW10 изменения в таблице

Внесла изменения в соответствии с комментариями.
На сколько я понимаю, неверно вызываю данные из массива, но не могу придумать как должно происходить. Сейчас выходит ошибка - "Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value".
Как-то с таблицами совсем не заладилось.

TatyanaV24 committed:
Update UserInterfaceState.xcuserstate

.

ios-review commented on HW10/HW10/SetTableViewCell.swift:
```diff
@@ -0,0 +1,27 @@
+        super.setSelected(selected, animated: animated)
+
+        // Configure the view for the selected state
+    }
```
пустые методы не нужны

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -7,56 +7,63 @@
         super.viewDidLoad()
-        
     }
 }
```
удалить эти строки

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -7,56 +7,63 @@
-        
     }
 }
 extension TableViewController: UITableViewDataSource, UITableViewDelegate {
```
extension SettingsViewController: ....

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -7,56 +7,63 @@
     
     func numberOfSections(in tableView: UITableView) -> Int {
-        return cont.count
+        return cont.settingSections.count
```
return settingsSections.count

ios-review reviewed (changes requested):
SettingsViewController и TableViewController разные классы.
Проще выбрать что-то одно.
И работать именно с этим классом и его данными.

Оставьте SettingsViewController.
<img width="1404" alt="Screenshot 2021-01-20 at 17 12 35" src="https://user-images.githubusercontent.com/62285448/105160469-dd145b00-5b42-11eb-9f59-49970df72453.png">

TatyanaV24 committed:
измнения в таблице

Я дополнила таблицу, она грузится практически верно.
У меня остались вопросы еще по двум элемента:
1. Для добавлению в строку "Wi-Fi" названия сети, должен создаваться еще лейбл, или есть функция, которая позволяет добавлять текст.
2. Также есть строка "Основные" на скрине есть оповещение ввиде красного кружка. Тут тоже должна быть функция, или он должен добавляться в ранее созданную структуру.
Спасибо.

ios-review commented:
SettingsViewController и TableViewController разные классы.
Это причина ошибки при запуске приложения.

TatyanaV24 committed:
.

Дополнения. Добавила еще задачу по UIScrollView

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -22,48 +22,53 @@ struct SettingSection {
+            cell.accessoryType = .none
+        } else {
+            cell.accessoryType = .disclosureIndicator
+        }
```
В этом методе не должно быть зависимости на конкретные значения indexPath.
Вынести accessoryType в модель.

ios-review commented on HW10/HW10/ScrollViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
+        return 0
+    }
```
удалить и все будет отображаться

ios-review reviewed (changes requested):
Комментарии + скрины
На скринах в консоли показаны ошибки layout.

<img width="1414" alt="Screenshot 2021-01-26 at 18 33 15" src="https://user-images.githubusercontent.com/62285448/105840195-32e67880-6005-11eb-8648-fa7df005d7c6.png">
<img width="1400" alt="Screenshot 2021-01-26 at 18 33 28" src="https://user-images.githubusercontent.com/62285448/105840241-45f94880-6005-11eb-958d-61764a20dbcb.png">
<img width="1401" alt="Screenshot 2021-01-26 at 18 33 46" src="https://user-images.githubusercontent.com/62285448/105840289-5dd0cc80-6005-11eb-8d83-94015fc15aba.png">

<img width="1402" alt="Screenshot 2021-01-26 at 18 30 43" src="https://user-images.githubusercontent.com/62285448/105840120-15b1aa00-6005-11eb-9fcb-49a30fd6877d.png">

TatyanaV24 committed:
.

TatyanaV24 committed:
HW and ScrollView

На сколько я понимаю, мне удалось исправить ошибки с Autolayout.
Осталась одна ошибка в TableViewController - HW10[13710:1407805] [Storyboard] Unknown class name in Interface Builder file. Не могу понять, какой именно класс он видит...а я нет.
Также переделала контроллер со ScrollView, к сожалению не могу понять, что именно в нем не так срабатывает. По идее при помощи ScrollView  должна меняться картинка с котом, а также прокручиваться весь экран, если не хватает место.
Также тут что-то не то с констрейтами, я не понимаю, почему в на 11 айфоне, все в порядке, на SE  идет обрезка. Старалась все задавать максимально без привязки к какому-то одному экрану.

ios-review commented on HW10/HW10/Controller/ViewController.swift:
```diff
@@ -0,0 +1,86 @@
+}
+
+extension ViewController: UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        let w = UIScreen.main.bounds.size.width / 2
+        return CGSize(width: w, height: w)
+    }
```
Зависимость размера ячейки от размеров экрана может сделать ячейку непропорциональной размерам коллекции, в случае если коллекция будет использоваться как часть большого экрана.

Рекомендую использовать зависимость от collectionView.bounds.

ios-review commented on HW10/HW10/Controller/ViewController.swift:
```diff
@@ -0,0 +1,86 @@
+    
+    @IBOutlet weak var collectioView: UICollectionView!
+    
+    var itemMenuArray:[Menu] = {
```
`let itemMenu`

ios-review commented on HW10/HW10/Model/Model.swift:
```diff
@@ -0,0 +1,16 @@
+    var discount: String?
+    var name: String?
+    var percent: String?
+}
```
```
struct Menu {
    let imageName: String
}
```

Использование константных свойств делает код безопаснее, так как меньшее возможности для изменения значений.

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -0,0 +1,76 @@
+            cell.accessoryType = .none
+        } else {
+            cell.accessoryType = .disclosureIndicator
+        }
```
в модели данных есть свойство accessoryType, но здесь не используется.
`cell.accessoryType = name.accessoryType`

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -0,0 +1,76 @@
+    let settings: [Setting]
+}
+
+class TableViewController: UIViewController {
```
В сториборде где-то указан класс, которого на самом деле в проекте нет.
<img width="1394" alt="Screenshot 2021-02-04 at 18 51 42" src="https://user-images.githubusercontent.com/62285448/106889813-1c35d500-671b-11eb-8e12-a639138ed581.png">

ios-review commented on HW10/HW10/SomeScrollViewController.swift:
```diff
@@ -0,0 +1,50 @@
+        
+        imageViewRect.origin.x += imageViewRect.size.width
+        let threeImage = self.newImageViewWithImage(paramImage: UIImage(named: "3")!, paramFrame: imageViewRect)
+        imgScroll.addSubview(threeImage)
```
Скролл вью более сложная вью для реализации layout.
Если из кода, то необходимо еще задать content size.

ios-review replied:
`imgScroll.contentSize = CGSize(width: 1000, height: imgScroll.bounds.size.height)`

ios-review replied:
Но в этом случае нельзя использовать констрейнты.
Если используются констрейнты, то contentSize будет вычисляться по их правилам.

ios-review reviewed (changes requested):
Задание выполнено в соответствии с требованиями👍

Попробуете доделать скрол вью в скрол вью?

TatyanaV24 committed:
правки 13.02

Все исправлено и работает.

ios-review commented on HW10/HW10/Controller/ViewController.swift:
```diff
@@ -11,53 +11,17 @@ class ViewController: UIViewController {
-        blankMenu6.percent = "- 50 %"
-        blankMenu6.price = "1 500 руб."
-        blankMenu6.name = "Котик малыш"
+    var itemMenu:[Menu] = {
```
`let itemMenu: [Menu]`

По умолчанию используйте константы.
Они безопаснее, так как не позволят другому программисту изменить их значение, а значит всегда можно полагаться на их значение.
Другой программист может изменить константу на переменную, но в случае использования констант по умолчанию, он подумает несколько раз прежде чем изменит на переменную.

itemMenu нигде в коде дальше не меняется.

Такая запись самая короткая.
```
let itemMenu = [
    Menu(imageName: "1", price: "5 000 руб.", discount: "3 500 руб", name: "Котик Тимоха в желтой толстовке", percent: "- 30 %"),
    Menu(imageName: "2", price: "4 300 руб.", discount: "6 500 руб", name: "Котик-сплюшка в колпаке", percent: "- 35 %")
]
```

ios-review commented on HW10/HW10/Model/Model.swift:
```diff
@@ -8,9 +8,9 @@
-    var discount: String?
-    var name: String?
-    var percent: String?
+    var imageName: String
```
Используйте константы!

ios-review replied:
let по умолчанию

ios-review commented on HW10/HW10/TableViewController.swift:
```diff
@@ -69,7 +69,7 @@ extension TableViewController: UITableViewDataSource, UITableViewDelegate {
         } else {
-            cell.accessoryType = .disclosureIndicator
+            cell.accessoryType = name.accessoryType
         }
```
Значение accessoryType необходимо использовать из модели для всех ячеек вне зависимости от значения index path.

ios-review replied:
`cell.accessoryType = name.accessoryType` вместо строк 69-73

ios-review reviewed (approved):
Все задания работают в полном соответствии с условиями 🔥

Реквест можно мержить.

Прошу обратить внимание на комментарии.
В следующих реализациях необходимо их учитывать.

TatyanaV24 committed:
.
